bots/doc.go

```diff
@@ -0,0 +1,139 @@
+with its designated third-party service, and the requests must be sent by
+an authorized CONIKS client via the service's protocol to the proxy's
+reserved account. As a result, a proxy trusts that any requests sent
+by CONIKS clients to its account originated from uncompromised
+third-party accounts.
+
+CONIKS servers: bots proxies relay verified CONIKS registration requests
+to a designated CONIKS server. The server and proxies may run as separate
+processes on the same machine, or on separate machines, depending on the
+communication service's infrastructure. For authentication, bots proxies
+can be configured to digitally sign all requests they relay, or to
+establish a secure network connection with the server. The server then
+only accepts incoming registration requests originating from authenticated
+account proxies.
+
+Privacy
+
+bots seeks to provide the same strong privacy guarantees as CONIKS.
+CONIKS clients using bots only store the generated key pairs and
+registered third-party usernames. bots only sends the username and
+the public key as an opaque data blob to the corresponding account proxy.
+bots also ensures that the clients do not gain access to any other
+third-party account information.
+
+The account proxies only have access to a user's username with their
+designated service and the received public key, and this is
+the only user information they send to the server.
+Proxies cannot access any other account information.
+
+Third-Party Account Verification Protocol
+
+1. The user registers her third-party account with her communication
+client, thereby authorizing the client and the integrated CONIKS client
+to connect to her third-party account.
+
+2. The CONIKS client generates a key pair for the new user account
+and stores the public and private key locally.
+
+3. On behalf of the connected account, the client sends a registration
+request
+    req = (username, opaque public key data blob)
+to the designated account proxy using the service's communication
+protocol.
+
+4. The proxy receives the request at its reserved account via
+the service's protocol, and verifies that the sender
+is an authorized client.
+
+5. The proxy sends the verified registration request to its designated
+CONIKS server.
+
+6. Once the CONIKS server receives the request and authenticates the proxy,
+it registers the username-to-public key mapping in its CONIKS directory.
+The server sends a cryptographic proof of registration to the proxy,
+which relays the response back to the CONIKS client.
+
+Supported Third-Party Services
+
+The current implementation supports third-party account verification of
+Twitter accounts. The account proxy connects to its reserved Twitter
+account using OAuth, and it only accepts CONIKS registration
+requests received as Twitter direct messages sent from CONIKS clients
+that have been authorized by a legitimate Twitter account.
+
+Limitations
```

Comment:
> - First party: keyserver is run by identity provider
- Third party: keyserver is run by communication service (or really anyone) on behalf of the identify provider

Ok, gotcha. I think, though, that the bot is mostly relevant in the third-party case, no? I'm not sure that a first-party would run a bot.

> verifying the ownership of the key is limited by the allowsPublicVisibility flag, and seems the same in the two cases.

Right.

>  I was talking about, "we avoid exposing the user's public key to the registration proxy" (which is similarly affected by that flag, but let's ignore that for now).

Ah I see, but the limitation I was pointing out wasn't the we avoid the exposure of the public key, but rather that "this also limits the proxy's ability to verify explicitly the ownership of this public key". But given the use of the `allowsPublicVisibility` flag, this might not be feasible anyway. I think the real reason why this is a limitation isn't so much that the data isn't exposed, but rather because the bot can't make any assumptions on what the key data looks like. 

> Is that even true though, we aren't encrypting the data?

What I meant by "avoid exposing the public key" is that the client only sends the fingerprint, but this is an implementation detail in the client and because of the fact that the bot doesn't make any assumptions about the key data, I don't think this point should be mentioned. But I don't think we're encrypting the data sent to the bot, unless https://github.com/c633/ctypes-otr/blob/coniks-registration/chrome/content/coniks/coniks.js#L207 encrypts the data. And AFAIK, the Twitter OAuth API only signs the requests.

>  In the third-party case (my terminology above), we have no proof that the identity provider requested the registration.  If we did, then we'd be able to determine who was being malicious. But also because I wasn't sure where else we had planned to write this up.

Ah, this is the difference between the first- and third-party registration you were saying I should describe? It's true, the bot doesn't generate cryptographic proof that the identity provider made the request (I think we had briefly talked about this at some point, about somehow using the TLS handshake as cryptographic proof of the identity provider's involvement). I think the registration bot documentation is a good place for writing this up since I think this is mostly problematic during registration with a third-party keyserver.